core/src/main/java/org/elasticsearch/index/mapper/Uid.java

```diff
@@ -123,4 +126,169 @@ public static String createUid(String type, String id) {
+        }
+        return new BytesRef(b);
+    }
+
+    /** With base64 ids, we decode and prepend an escape char in the cases that
+     *  it could be mixed up with numeric or utf8 encoding. In the majority of
+     *  cases (253/256) the encoded id is exactly the binary form. */
+    private static BytesRef encodeBase64Id(String id) {
+        byte[] b = Base64.getUrlDecoder().decode(id);
+        if (Byte.toUnsignedInt(b[0]) >= BASE64_ESCAPE) {
+            byte[] newB = new byte[b.length + 1];
+            newB[0] = (byte) BASE64_ESCAPE;
+            System.arraycopy(b, 0, newB, 1, b.length);
+            b = newB;
+        }
+        return new BytesRef(b, 0, b.length);
+    }
+
+    private static BytesRef encodeUtf8Id(String id) {
+        byte[] b = new byte[1 + UnicodeUtil.maxUTF8Length(id.length())];
+        // Prepend a byte that indicates that the content is an utf8 string
+        b[0] = (byte) UTF8;
+        int length = UnicodeUtil.UTF16toUTF8(id, 0, id.length(), b, 1);
+        return new BytesRef(b, 0, length);
+    }
+
+    /** Encode an id for storage in the index. This encoding is optimized for
+     *  numeric and base64 ids, which are encoded in a much denser way than
+     *  what UTF8 would do.
+     *  @see #decodeId */
+    public static BytesRef encodeId(String id) {
+        if (id.isEmpty()) {
+            throw new IllegalArgumentException("Ids can't be empty");
+        }
+        if (isPositiveNumeric(id)) {
+            // common for ids that come from databases with auto-increments
+            return encodeNumericId(id);
+        } else if (isURLBase64WithoutPadding(id)) {
+            // common since it applies to autogenerated ids
+            return encodeBase64Id(id);
+        } else {
+            return encodeUtf8Id(id);
+        }
+    }
+
+    private static String decodeNumericId(byte[] idBytes) {
+        assert Byte.toUnsignedInt(idBytes[0]) == NUMERIC;
+        int length = (idBytes.length - 1) * 2;
+        char[] chars = new char[length];
+        for (int i = 1; i < idBytes.length; ++i) {
+            final int b = Byte.toUnsignedInt(idBytes[i]);
+            final int b1 = (b >>> 4);
+            final int b2 = b & 0x0f;
+            chars[(i - 1) * 2] = (char) (b1 + '0');
+            if (i == idBytes.length - 1 && b2 == 0x0f) {
+                length--;
+                break;
+            }
+            chars[(i - 1) * 2 + 1] = (char) (b2 + '0');
+        }
+        return new String(chars, 0, length);
+    }
+
+    private static String decodeUtf8Id(byte[] idBytes) {
+        assert Byte.toUnsignedInt(idBytes[0]) == UTF8;
+        return new BytesRef(idBytes, 1, idBytes.length - 1).utf8ToString();
+    }
+
+    private static String decodeBase64Id(byte[] idBytes) {
+        assert Byte.toUnsignedInt(idBytes[0]) <= BASE64_ESCAPE;
+        if (Byte.toUnsignedInt(idBytes[0]) == BASE64_ESCAPE) {
+            idBytes = Arrays.copyOfRange(idBytes, 1, idBytes.length);
```

Comment:
It is unfortunate `Base64.Encoder.encodeToString` only has a `byte[]` variant and not a `ByteBuffer` version that would allow avoiding the copy...